packages/wrangler/src/parse.ts

```diff
@@ -0,0 +1,178 @@
+import { readFile as readFile_ } from "node:fs/promises";
+import TOML from "@iarna/toml";
+import { formatMessagesSync } from "esbuild";
+import type { PartialMessage, Location } from "esbuild";
+
+/**
+ * Formats a message using esbuild's pretty-print algorithm.
+ */
+export function formatMessage(
+  message: PartialMessage,
+  kind: "warning" | "error" = "error",
+  label?: string
+): string {
+  const lines = formatMessagesSync([message], {
+    color: true,
+    kind: kind,
+    terminalWidth: process.stderr.columns,
+  });
+  if (label) {
+    lines[0] = lines[0].replace(
+      kind.toUpperCase(),
+      `${label.toUpperCase()} ${kind.toUpperCase()}`
+    );
+  }
+  return lines.join("\n");
+}
+
+/**
+ * Formats and prints a message to stderr.
+ */
+export function printMessage(
+  message: PartialMessage,
+  kind?: "warning" | "error",
+  label?: string
+): void {
+  const text = formatMessage(message, kind, label);
+  process.stderr.write(text);
+}
+
+/**
+ * An error that's thrown when something fails to parse.
+ */
+export class ParseError extends Error {
+  readonly detail: PartialMessage;
+
+  constructor(message: PartialMessage) {
+    super(message.text);
+    this.name = this.constructor.name;
+    this.detail = message;
+    if (!this.detail.notes) {
+      this.detail.notes = [];
+    }
```

Comment:
Be aware that you are mutating the original `message` that is passed in here.
Perhaps a safer way would be to make a shallow clone?

```ts
this.detail = { notes: [], ...message };
```